File: backend/src/main/java/me/kzv/okvue/config/SecurityConfig.java
package me.kzv.okvue.config;

import lombok.RequiredArgsConstructor;
import me.kzv.okvue.security.oauth2.CustomOAuth2UserService;
import me.kzv.okvue.security.oauth2.handler.CustomOAuth2FailureHandler;
import me.kzv.okvue.security.oauth2.handler.CustomOAuth2SuccessHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;

@RequiredArgsConstructor
@EnableWebSecurity
@Configuration
public class SecurityConfig {

    private final CustomOAuth2UserService customOAuth2UserService;
    private final CustomOAuth2SuccessHandler successHandler;
    private final CustomOAuth2FailureHandler failureHandler;

    @Bean
    public BCryptPasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {

        http
                .csrf().disable()
                .headers().frameOptions().disable()
                .and()
                .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS)
                .and()
                .authorizeRequests()
                // h2-console설정
                .antMatchers("/h2-console/**").permitAll()
                .antMatchers("/", "/login", "/admin/login", "/api/v1/login").permitAll()
//                .antMatchers("/api/v1/**").hasRole(Role.USER.name())
//                .anyRequest().authenticated()
                .antMatchers("/admin/**").authenticated()
                .anyRequest().permitAll()
                .and()
                .formLogin().loginPage("/admin/login").permitAll() // 어드민페이지 로그인 페이지
                .and()
                .logout().logoutSuccessUrl("/admin").permitAll()
                .and()
                .oauth2Login()
                .loginPage("/login") // 스프링 기본 로그인 페이지 사용 안하려면 매핑해줘야함
                .userInfoEndpoint()
                .userService(customOAuth2UserService)
                .and()
                .successHandler(successHandler)
                .failureHandler(failureHandler)
        ;

//        http.headers().frameOptions().sameOrigin();

        return http.build();
    }

    @Bean
    public WebSecurityCustomizer webSecurityCustomizer() {
        return (web) -> web.ignoring().antMatchers("/images/**", "/js/**", "/webjars/**", "/favicon.ico");
    }
}

File: rest/backend/src/main/kotlin/me/kzv/board/config/SecurityConfig.kt
package me.kzv.board.config

import org.springframework.boot.autoconfigure.security.servlet.PathRequest
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.builders.WebSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain

@Configuration
@EnableWebSecurity
class SecurityConfig(

) {

    @Bean
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .csrf().disable()
            .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS)
            .and()
            .headers().frameOptions().disable()
            .and()

            // url 권한 설정
            .authorizeRequests()
            .antMatchers("/h2-console/**")
            .permitAll()

            .antMatchers("/", "/api/v1/signin", "/api/v1/signup")
            .permitAll()
            .antMatchers("/api/v1/member/**").hasRole("member")
            .antMatchers("/api/v1/admin/**").hasRole("ADMIN")
            .anyRequest().authenticated()

            // 기본 인증 로그인 disable
            .and()
            .formLogin().disable()
            .logout().disable()
            .httpBasic().disable()


        return http.build()
    }

    @Bean
    fun webSecurityCustomizer(): WebSecurityCustomizer = WebSecurityCustomizer { web: WebSecurity ->
        web.ignoring().requestMatchers(PathRequest.toStaticResources().atCommonLocations())
//        web.ignoring().mvcMatchers("/vendor/**")
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder = BCryptPasswordEncoder()
}